Refactor/Fix sponge function

moCello opened the pull request:
I used the change in the function signatures for issue #50 to also refactor sponge internally to make it easier to follow.

The capacity of the state was initialized with a one but it must be a zero -> Fixed

We used to have sponge return an output vector of variable length. However in our use case this output vector will always be one element long. We therefore decided to remove complexity and get rid of the `output_len` parameter. We will always return one `BlsScalar`.
See also the discussion bellow

moCello committed:
Allign calling convention `sponge` and `permutation`

Refactor sponge internally for readability.
Fix the initialization of the capacity in the state of sponge:
It was one and is now zero.

Resolves: #50

vlopes11 commented on src/sponge.rs:
```diff
@@ -8,131 +8,120 @@
-    // Initialize input as mutable
-    let mut input = input;
-
+pub fn hash(input: &[BlsScalar], output_len: usize) -> Vec<BlsScalar> {
```
There is something off with this function. Normally we use a single bit digest as output:

`pub fn hash(input: &[BlsScalar]) -> BlsScalar`

Here we are mixing functionalities. You are creating an iterator that could squeeze `n` elements, while also providing a normal digest function.

I suggest you split this function in two:

`pub fn digest(input: &[BlsScalar]) -> BlsScalar`
`pub fn stream(input: &[BlsScalar]) -> impl Iterator<Item = BlsScalar>`

vlopes11 replied:
For that, you will need some iterator wrapper struct that will contain a state and its used element

moCello replied:
Not sure I understand what you mean. I follow the sponge framework that is divided in three phases:
- Pad `input`
- Absorb `input` into `state`
- Squeeze `state` to generate `output` of the given length `output_len`

I could of course create three different functions for those phases but I don't really see what is being achieved with that. But maybe I am missing your point.

moCello replied:
We talked earlier about the output of sponge. It actually can be of different length which is why we have `output_len` as a parameter. If we decide however that we want ours to always return one `BlsScalar` I could of course change that.

vlopes11 replied:
We are talking specifically about the `squeeze` phase. The `pad` and `absorb` is fine

Yes, the output might be variable size. However, many use cases will only get one field element; so its a waste to create a vector to store a single element.

As for the variable size, if we are creating a vector, we are eager loading and allocating unnecessarily. If its an iterator, the consumer can decide what to do with that (including creating a vector with `collect::<Vec<BlsScalar>>`

moCello replied:
Ok, I see. I think you are right then and the easiest would be to simply return one `BlsScalar` so we don't introduce complexity that we don't use anyway.

vlopes11 replied:
We might use multiple outputs of the sponge tho. We could leave an open issue to implement the iterator in the future

moCello replied:
Sure! Although given the fact that we didn't need sponge to return multiple outputs up until this point, I find it unlikely that we will need it in the future. But one never knows.

vlopes11 commented on src/sponge.rs:
```diff
@@ -8,131 +8,120 @@
-    // Initialize input as mutable
-    let mut input = input;
-
+pub fn hash(input: &[BlsScalar], output_len: usize) -> Vec<BlsScalar> {
```
There is something off with this function. Normally we use a single bit digest as output:

`pub fn hash(input: &[BlsScalar]) -> BlsScalar`

Here we are mixing functionalities. You are creating an iterator that could squeeze `n` elements, while also providing a normal digest function.

I suggest you split this function in two:

`pub fn digest(input: &[BlsScalar]) -> BlsScalar`
`pub fn stream(input: &[BlsScalar]) -> impl Iterator<Item = BlsScalar>`

moCello committed:
Change sponge to always return one scalar

In out usage of `sponge` the `output_len` will always set to `1`, so we
can discard the squeezing phase and simply return the first element of
`state`.

See coomments in associated PR #63 for the discussion.

moCello committed:
Update CHANGELOG

vlopes11 commented on src/sponge.rs:
```diff
@@ -4,135 +4,65 @@
+pub fn hash(input: &[BlsScalar]) -> BlsScalar {
     // Pad input
+    // After this, input_len is a multiple of 2
+    let mut input = input.to_vec();
```
As improvement for the future (or now), this can easily be an iterator instead of a vector

moCello replied:
You are right, I'll change it now

vlopes11 commented on tests/sponge.rs:
```diff
@@ -12,9 +12,9 @@ use zelbet::sponge;
 
 fn sponge_circuit(composer: &mut TurboComposer, input: BlsScalar) {
```
Also for the future, but it doesn't really make sense to test a sponge function with a single scalar. This test should cover edge cases such as 0, 1, 2 and many scalars as input

moCello replied:
Good point! I will add more test cases

vlopes11 reviewed (approved):
LGTM

moCello committed:
Use iterator instead of vector

This was raised by vlopes11 in the review.
This commit also expands the sponge test cases to more different
input lengths.

ureeves reviewed (approved):
LGTM

moCello committed:
Update function signtures in benchmarks

vlopes11 reviewed (approved):
LGTM
